File: app/javascript/hyacinth_v1/graphql/dynamicFields.js
import gql from 'graphql-tag';

export const getDynamicFieldQuery = gql`
  query DynamicField($id: ID!){
    dynamicField(id: $id) {
      id
      stringKey
      displayLabel
      sortOrder
      controlledVocabulary
      fieldType
      filterLabel
      isFacetable
      isIdentifierSearchable
      isKeywordSearchable
      isTitleSearchable
      selectOptions
    }
  }
`;

export const dynamicFieldPathQuery = gql`
  query DynamicFieldPathQuery($id: ID!) {
    dynamicField(id: $id) {
      id
      displayLabel
      type: __typename
      __typename # Apollo can't correctly cache if we override this field. Eventually we might override how cache identifiers are generated.
      path {
        ...on DynamicFieldGroup {
          id
          displayLabel
        }
        ...on DynamicFieldCategory {
          id
          displayLabel
        }
        type: __typename
        __typename
      }
    }
  }
`;

export const createDynamicFieldMutation = gql`
  mutation CreateDynamicField($input: CreateDynamicFieldInput!) {
    createDynamicField(input: $input) {
      dynamicField {
        id
      }
    }
  }
`;

export const updateDynamicFieldMutation = gql`
  mutation UpdateDynamicField($input: UpdateDynamicFieldInput!) {
    updateDynamicField(input: $input) {
      dynamicField {
        id
      }
    }
  }
`;

export const deleteDynamicFieldMutation = gql`
  mutation DeleteDynamicField($input: DeleteDynamicFieldInput!) {
    deleteDynamicField(input: $input) {
      dynamicField {
        id
      }
    }
  }
`;

File: spec/factories/dynamic_fields.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :dynamic_field do
    string_key      { 'term' }
    display_label   { 'Value' }
    field_type      { DynamicField::Type::CONTROLLED_TERM }
    sort_order      { 7 }
    filter_label    { 'Name' }

    is_facetable             { true }
    controlled_vocabulary    { 'name_role' }
    select_options           { nil }
    is_keyword_searchable    { false }
    is_title_searchable      { false }
    is_identifier_searchable { false }

    created_by      { User.first || create(:user) }
    updated_by      { User.first || create(:user) }

    association :dynamic_field_group, factory: :dynamic_field_group, strategy: :create

    trait :string do
      field_type { DynamicField::Type::STRING }
      controlled_vocabulary { nil }
    end
  end
end
